DataFormats/TrackReco/interface/TrackExtraBase.h

```diff
@@ -51,8 +51,13 @@ class TrackExtraBase
     }
 
     /// get a ref to i-th recHit
-    TrackingRecHitRef recHitRef(unsigned int i) const {                                                               
-        return TrackingRecHitRef(m_hitCollection,m_firstHit+i);
+    TrackingRecHitRef recHitRef(unsigned int i) const {
+      if(m_hitCollection.productPtr()) {
+        TrackingRecHitRef::finder_type finder;
+        TrackingRecHitRef::value_type const* item = finder(*(static_cast<TrackingRecHitRef::product_type const*>(m_hitCollection.productPtr())), m_firstHit+i);
+        return TrackingRecHitRef(m_hitCollection.id(), item, m_firstHit+i);
+      }
+      return TrackingRecHitRef(m_hitCollection,m_firstHit+i);

```

Comment:
The Ref contains a transient pointer as a data member. Sometimes it points at the element (in earlier versions of the code it used to point at the product), but when the Ref has just been read from an input file (and in some other circumstances) the pointer points at an EDProductGetter (a pointer which is set in a custom streamer). The EDProductGetter can get the product from the Event using the ProductID that the Ref also contains. That part of Ref has not changed.  If the function productPtr returns null, then that means the pointer in the RefCore is pointing at the EDProductGetter. And a Ref in that state can be constructed from just a RefCore and a key. In that state, RefProd and Ref contain the same pointer to the EDProductGetter. Somewhat complicated. Ask again if I did not explain it clearly enough.